controllers/control/k8ssandratask_controller.go

```
@@ -0,0 +1,212 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/k8ssandra/k8ssandra-operator/apis/control/v1alpha1"
+)
+
+// K8ssandraTaskReconciler reconciles a K8ssandraTask object
+type K8ssandraTaskReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme      *runtime.Scheme
+	ClientCache *clientcache.ClientCache
+}
+
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/finalizers,verbs=update
+
+func (r *K8ssandraTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("K8ssandraTask", req.NamespacedName)
+
+	logger.Info("Fetching task", "K8ssandraTask", req.NamespacedName)
+	k8Task := &api.K8ssandraTask{}
+	err := r.Get(ctx, req.NamespacedName, k8Task)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: r.ReconcilerConfig.DefaultDelay}, err
+	}
+	// TODO k8Task.DeepCopy()? why?
+
+	if k8Task.DeletionTimestamp != nil {
+		// TODO delete dependent objects?
+		return ctrl.Result{}, nil
+	}
+
+	kcKey := k8Task.GetClusterKey()
+	logger.Info("Fetching cluster", "K8ssandraCluster", kcKey)
+	kc := &k8sapi.K8ssandraCluster{}
+	err = r.Get(ctx, kcKey, kc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	dcs, err := filterDcs(kc, k8Task.Spec.Datacenters)
+	if err != nil {
+		return ctrl.Result{}, err
```

Do we wish to entertain user with the knowledge that their spec is incorrect?

---

We should surface the error in the status in some way. Consider trying to build some automation around this such that we should get notification if the tasks fails or is unable to run. We would grep the operator logs, but that would be brittle. It would easier and better to be able to query the status.

---

But we do not tell the user where that DC is wrong. Lets say in rebuild, was it incorrect in the source_datacenter or target datacenters?

---

I've since switched to Kubernetes events for user error reporting (see the docs added by this PR).